NXP-29516/NXP-30044: async blob digest computation (10.10)

efge opened the pull request.

nuxeojenkins commented:
View issues in JIRA:
- [NXP-29516](https://jira.nuxeo.com/browse/NXP-29516): Allow efficient search by blob key
- [NXP-30044](https://jira.nuxeo.com/browse/NXP-30044): Async blob digest computation

kevinleturc commented on addons/nuxeo-core-binarymanager-cloud/nuxeo-core-binarymanager-s3/src/test/java/org/nuxeo/ecm/blob/s3/TestS3BlobStoreTracing.java:
```diff
@@ -0,0 +1,522 @@
+        try {
+            LogTracingHelper.assertEqualsLists(expectedTrace, actualTrace, context);
+        } catch (AssertionError e) {
+            System.err.println(filename);
+            System.err.println(String.join("\n", actualTrace));
```
To replace by a logger?

efge replied:
It's only for when the assertion fails, it's easier for me to have it on stderr.

kevinleturc commented on nuxeo-core/nuxeo-core-api/src/main/java/org/nuxeo/ecm/core/blob/AbstractBlobStore.java:
```diff
@@ -101,6 +115,76 @@ public String writeBlob(BlobContext blobContext) throws IOException {
+     *
+     * @since 11.5
+     */
+    protected abstract String writeBlobGeneric(BlobWriteContext blobWriteContext) throws IOException;
```
Should we put a default implementation for the backport?

efge replied:
The new default implem `AbstractBlobStore.writeBlob(BlobWriteContext)` is the only one that calls `writeBlobGeneric`. And an old customer `AbstractBlobStore` subclass had to have an existing `writeBlob(BlobWriteContext)` (because it's in the interface), so any existing customer implementation will continue to work and never call `writeBlobGeneric`. So I think we're ok.

efge replied:
Hm OTOH it's `abstract` so recompilation would fail... Ok I'll add a default implementation throwing something.

kevinleturc commented on nuxeo-core/nuxeo-core-test/src/test/java/org/nuxeo/ecm/core/blob/TestComputeDigestHelper.java:
```diff
@@ -0,0 +1,151 @@
+        session.save();
+
+        // make doc1 point to doc2's blob
+        Listener.properties = null;
+        String oldDigest = session.replaceBlobDigest(doc1.getRef(), FOO_MD5, BAR_MD5, "newdigest");
+        assertEquals(FOO_MD5, oldDigest);
+        // check event
+        assertNotNull(Listener.properties);
+        assertEquals(FOO_MD5, Listener.properties.get("key"));
+        assertEquals(FOO_MD5, Listener.properties.get("digest"));
+        assertEquals(BAR_MD5, Listener.properties.get("newKey"));
+        assertEquals("newdigest", Listener.properties.get("newDigest"));
```
Maybe using `CapturingEventListener` here?

efge replied:
Good idea

kevinleturc reviewed (approved):
Great work 👍.

efge commented:
Updated previous commits to answer comments.
Updated previous commit to add a `maxThreads` to the `ComputeDigestWork` queue config.
Added a new commit for NXP-30043 for safe concurrent behavior (using kv store).

efge commented:
A few fixups in previous commits:
- log the new event `blobDigestUpdated` in the audit with the proper extended info (and CoreEventConstants for the keys).
- added `digestAlgorithm` with getter/setter in `AbstractBlob`, and updated `BlobStoreBlobProvider.fixupDigest` to set it correctly (this is tested in `TestS3BlobStoreTracing.testCopyAsyncDigest`).

Added a commit for NXP-30119 for the Drive Server changes.

efge commented:
(last force push is just a 10.10 rebase)

efge commented:
T&P ok https://qa2.nuxeo.org/jenkins/job/TestAndPush/job/ondemand-testandpush-fguillaume-10.10/142/

efge commented:
(rebased)

efge commented:
(fixed one Javadoc issue + rebased)

aimiebell commented on addons/nuxeo-core-binarymanager-cloud/nuxeo-core-binarymanager-s3/src/test/java/org/nuxeo/ecm/blob/s3/TestS3BlobStoreNoCaching.java:
```diff
@@ -37,4 +38,9 @@ public void testFlags() {
 
+    @Override
+    public void testBlobGetFile() {
+        new AssumptionViolatedException("S3BlobStore with no caching has no File");
```
I detect that this code is problematic. According to the [Correctness (CORRECTNESS)](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#correctness-correctness), [RV: Exception created and dropped rather than thrown (RV_EXCEPTION_NOT_THROWN)](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#rv-exception-created-and-dropped-rather-than-thrown-rv-exception-not-thrown).
This code creates an exception (or error) object, but doesn't do anything with it. For example, something like
It was probably the intent of the programmer to throw the created exception:

efge replied:
@aimiebell thanks. Please indicate in your message that this is a bot, it's bad form hide this info.

efge commented:
Added two commits to make path strategies accept more keys, and improve CachingBlobStore to better use path strategies (don't depend on LRUFileCache); with added caching-related tests.

efge commented:
T&P ok https://qa2.nuxeo.org/jenkins/job/TestAndPush/job/ondemand-testandpush-fguillaume-10.10/144/

efge commented:
New commit to disallow async digest on VCS.

efge commented:
Last force pushes were all about tweaking unit tests corner cases.

efge commented:
T&P ok https://qa2.nuxeo.org/jenkins/job/TestAndPush/job/ondemand-testandpush-fguillaume-10.10/150/

efge commented:
T&P ok https://qa2.nuxeo.org/jenkins/job/TestAndPush/job/ondemand-testandpush-fguillaume-10.10/152/
